core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -87,11 +105,44 @@ public static CloseableIterable<CombinedScanTask> planTasks(
                     + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
                 (1 + file.deletes().size()) * openFileCost);
 
-    return CloseableIterable.transform(
-        CloseableIterable.combine(
-            new BinPacking.PackingIterable<>(splitFiles, splitSize, lookback, weightFunc, true),
-            splitFiles),
-        BaseCombinedScanTask::new);
+    if (preservedPartitionIds != null && !preservedPartitionIds.isEmpty()) {
+      Preconditions.checkArgument(
+          spec != null, "spec can't be null when " + "preservedPartitionIds is not null");
+      StructProjection projectedStruct =
+          StructProjection.create(spec.partitionType(), preservedPartitionIds);
+      Types.StructType projectedPartitionType = projectedStruct.type();
+      ListMultimap<StructLikeWrapper, FileScanTask> groupedFiles =
+          Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
+
+      splitFiles.forEach(
+          f -> {
+            StructLikeWrapper wrapper =
+                StructLikeWrapper.forType(projectedPartitionType)
+                    .set(projectedStruct.copy().wrap(f.file().partition()));
+            groupedFiles.put(wrapper, f);
+          });
+
+      List<Iterable<BaseCombinedScanTask>> groupedTasks =
+          groupedFiles.asMap().values().stream()
+              .map(
+                  t ->
+                      Iterables.transform(
+                          new BinPacking.PackingIterable<>(
+                              CloseableIterable.withNoopClose(t),
```

Comment:
Since we hold these entries in memory, shall we also sort them by partition before combining?

Reply:
To pack files with the same partitions together. We will look back only at a small number of bins (configurable).

Reply:
Suppose we preserve boundaries for `p1` but our partition spec is `p1`, `p2`, `p3`. It may be beneficial to combine files with the exact partition `p1`, `p2`, `p3` rather than simply with the same `p1`.